File: app/src/main/java/ru/org/adons/cblock/service/BlockService.java
package ru.org.adons.cblock.service;

import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.media.AudioManager;
import android.os.IBinder;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.TaskStackBuilder;
import android.telephony.PhoneStateListener;
import android.telephony.TelephonyManager;

import com.android.internal.telephony.ITelephony;

import java.lang.reflect.Method;
import java.util.HashSet;
import java.util.Set;

import javax.inject.Inject;

import ru.org.adons.cblock.R;
import ru.org.adons.cblock.app.BlockManager;
import ru.org.adons.cblock.app.CBlockApplication;
import ru.org.adons.cblock.data.BlockListModel;
import ru.org.adons.cblock.ui.activity.MainActivity;
import ru.org.adons.cblock.utils.Logging;
import ru.org.adons.cblock.utils.SubscriptionUtils;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subscriptions.CompositeSubscription;

public class BlockService extends Service {

    private static final int NOTIFICATION_ID = 201507;
    private static volatile boolean isEnabled = false;

    private final CompositeSubscription subscription = new CompositeSubscription();
    private final Set<String> phones = new HashSet<>();

    private StateListener listener;
    private TelephonyManager manager;
    private ITelephony telephony;
    private AudioManager audio;

    @Inject BlockListModel blockListModel;
    @Inject BlockManager blockManager;

    /**
     * Start Block Service in foreground mode
     */
    public static synchronized void start(Context context) {
        if (!isEnabled) {
            isEnabled = true;
            context.startService(new Intent(context, BlockService.class));
        }
    }

    /**
     * Stop Block Service
     */
    public static synchronized void stop(Context context) {
        isEnabled = false;
        context.stopService(new Intent(context, BlockService.class));
    }

    @Override
    public void onCreate() {
        super.onCreate();
        // inject dependencies
        CBlockApplication application = (CBlockApplication) getApplication();
        ServiceComponent serviceComponent = DaggerServiceComponent.builder()
                .applicationComponent(application.applicationComponent())
                .build();
        serviceComponent.inject(this);

        // get blocked phones list
        Subscription dataSubscription = blockListModel.getBlockedPhones()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setPhones);
        subscription.add(dataSubscription);

        // subscribe to block list update from BlockManager
        Subscription updateSubscription = blockManager.getBlockListUpdate()
                .map(blockListModel::getBlockedPhones)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setPhones);
        subscription.add(updateSubscription);

        // register Phone State Listener
        listener = new StateListener();
        manager = (TelephonyManager) getSystemService(Context.TELEPHONY_SERVICE);
        manager.listen(listener, PhoneStateListener.LISTEN_CALL_STATE);
        try {
            Class<TelephonyManager> c = TelephonyManager.class;
            Method m = c.getDeclaredMethod("getITelephony");
            m.setAccessible(true);
            telephony = (ITelephony) m.invoke(manager);
        } catch (Exception e) {
            Logging.d(e.getMessage());
        }

        // init AudioManager
        audio = (AudioManager) getApplicationContext().getSystemService(Context.AUDIO_SERVICE);
    }

    private class StateListener extends PhoneStateListener {
        @Override
        public void onCallStateChanged(int state, String incomingNumber) {
            super.onCallStateChanged(state, incomingNumber);
            switch (state) {
                case TelephonyManager.CALL_STATE_RINGING:
                    if (phones.contains(incomingNumber)) {
                        // mute
                        int ringState = audio.getRingerMode();
                        boolean isMuted = false;
                        if (ringState != AudioManager.RINGER_MODE_SILENT) {
                            try {
                                audio.setRingerMode(AudioManager.RINGER_MODE_SILENT);
                                isMuted = true;
                            } catch (Exception e) {
                                Logging.d(e.getMessage());
                            }
                        }

                        // try to end call
                        if (telephony != null) {
                            try {
                                telephony.endCall();
                            } catch (Exception e) {
                                Logging.d(e.getMessage());
                            }
                        }

                        // un-mute
                        if (isMuted) {
                            audio.setRingerMode(ringState);
                        }
                        Logging.d("blocked number:" + incomingNumber);
                    }
                    break;
                case TelephonyManager.CALL_STATE_OFFHOOK:
                    break;
                case TelephonyManager.CALL_STATE_IDLE:
                    break;
                default:
                    break;
            }
        }
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        // set Notification - prevent service from stop by system
        NotificationCompat.Builder mBuilder = new NotificationCompat.Builder(this)
                .setSmallIcon(R.drawable.ic_notify)
                .setContentTitle(getApplicationContext().getString(R.string.main_notification_title))
                .setContentText(getApplicationContext().getString(R.string.main_notification_text_enable))
                .setAutoCancel(true)
                .setPriority(NotificationCompat.PRIORITY_MIN);

        // handle notification click
        Intent resultIntent = new Intent(this, MainActivity.class);
        TaskStackBuilder stackBuilder = TaskStackBuilder.create(this);
        stackBuilder.addParentStack(MainActivity.class);
        stackBuilder.addNextIntent(resultIntent);
        PendingIntent resultPendingIntent = stackBuilder.getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT);
        mBuilder.setContentIntent(resultPendingIntent);

        // start notification
        startForeground(NOTIFICATION_ID, mBuilder.build());
        return Service.START_STICKY;
    }

    @Override
    public void onDestroy() {
        // unsubscribe from all data update
        SubscriptionUtils.unsubscribe(subscription);

        // unregister Phone State Listener
        manager.listen(listener, PhoneStateListener.LISTEN_NONE);

        // cancel notification
        NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        notificationManager.cancel(NOTIFICATION_ID);

        // stop service
        synchronized (this) {
            isEnabled = false;
        }
        super.onDestroy();
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private void setPhones(Set<String> values) {
        phones.clear();
        phones.addAll(values);
    }

}

File: app/src/main/java/mobile/addons/cblock/service/BlockService.kt
package mobile.addons.cblock.service

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.IBinder
import android.telephony.PhoneStateListener
import android.telephony.TelephonyManager
import mobile.addons.cblock.app.BlockManager
import mobile.addons.cblock.app.CBlockApplication
import mobile.addons.cblock.ext.hideNotification
import mobile.addons.cblock.ext.safeUnsubscribe
import mobile.addons.cblock.ext.showNotification
import mobile.addons.cblock.model.BlockListModel
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import java.util.*
import javax.inject.Inject

class BlockService : Service() {

    companion object {
        @Volatile private var isEnabled = false

        // Start Block Service in foreground mode
        @Synchronized fun start(context: Context) {
            if (!isEnabled) {
                isEnabled = true
                context.startService(Intent(context, BlockService::class.java))
            }
        }

        // Stop Block Service
        @Synchronized fun stop(context: Context) {
            isEnabled = false
            context.stopService(Intent(context, BlockService::class.java))
        }
    }

    private val subscription = CompositeSubscription()
    private val phones = HashSet<String>()

    private lateinit var listener: StateListener
    private lateinit var manager: TelephonyManager

    @Inject lateinit var blockListModel: BlockListModel
    @Inject lateinit var blockManager: BlockManager

    override fun onCreate() {
        super.onCreate()
        initComponents()
        getData()
        getDataUpdate()
        setPhoneListener()
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        showNotification()
        return Service.START_STICKY
    }

    override fun onDestroy() {
        subscription.safeUnsubscribe()
        manager.listen(listener, PhoneStateListener.LISTEN_NONE)
        hideNotification()
        synchronized(this) { isEnabled = false }
        super.onDestroy()
    }

    override fun onBind(intent: Intent): IBinder? = null

    private fun setPhoneListener() {
        manager = getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
        listener = StateListener(applicationContext, manager, phones)
        manager.listen(listener, PhoneStateListener.LISTEN_CALL_STATE)
    }

    private fun initComponents() {
        val application = application as CBlockApplication
        val serviceComponent = DaggerServiceComponent.builder()
                .applicationComponent(application.applicationComponent)
                .build()
        serviceComponent.inject(this)
    }

    // get blocked phones list and
    private fun getData() {
        val dataSubscription = blockListModel.getBlockedPhones()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setPhones)
        subscription.add(dataSubscription)
    }

    // subscribe to block list update from BlockManager
    private fun getDataUpdate() {
        val updateSubscription = blockManager.getBlockListUpdate()
                .map { blockListModel.getBlockedPhones(it) }
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setPhones)
        subscription.add(updateSubscription)
    }

    private fun setPhones(values: Set<String>) {
        phones.clear()
        phones.addAll(values)
    }

}
